guides/building-a-middleware-from-scratch.md

```diff
@@ -0,0 +1,693 @@
+# Building a middleware from scratch
+
+In ["Inventing the `Service` trait"][invent] we learned all the motivations
+behind [`Service`] and why its designed the way it is. We also built a few
+smaller middleware ourselves but we took a few shortcuts in our implementation.
+In this guide we're going to build the `Timeout` middleware as it exists in
+Tower today without taking any shortcuts.
+
+Writing a robust middleware requires working with async Rust at a slightly lower
+level than you might be used to. The goal of this guide is to demystify the
+concepts and patterns so you can start writing your own middleware and maybe
+even contribute back to the Tower ecosystem!
+
+## Getting started
+
+The middleware we're going to build is [`tower::timeout::Timeout`]. It will set
+a limit on the maximum duration its inner `Service` is allowed to take. If it
+doesn't produce a response within some amount of time, an error is returned.
+This allows the client to retry that request or report an error to the user,
+rather than waiting forever.
+
+Lets start by writing a `Timeout` struct that holds the `Service` its wrapping
+and the duration of the timeout:
+
+```rust
+use std::time::Duration;
+
+struct Timeout<T> {
+    inner: T,
+    timeout: Duration,
+}
+```
+
+As we learned in ["Inventing the `Service` trait"][invent] its important for
```

Comment:
Hmm Im not sure if this is important it is just one method. The key point is that there are sometimes things you need to own for the response, ie the duration of the timeout but I believe even in timeout you can create the timeout future then move that in. So why would the service need to be clone (besides that if it can impl clone it should to forward). So I guess my point is that I think this circumvents why clone is important for service middleware rather than using it for moving the data it should be there to forward (and sometimes you do need to move self but not always).

Reply:
Yes, `Timeout` doesn't require its service to be Clone but I still think the general recommendation of making services Clone, so services can be moved into response futures, is important. So I think I'm going to leave this as is.